src/test/fuzz/util.h

```diff
@@ -534,4 +544,121 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
     }
 }
 
+class FuzzedSock : public Sock
+{
+    FuzzedDataProvider& m_fuzzed_data_provider;
+
+public:
+    explicit FuzzedSock(FuzzedDataProvider& fuzzed_data_provider) : m_fuzzed_data_provider{fuzzed_data_provider}
+    {
+    }
+
+    ~FuzzedSock() override
+    {
+    }
+
+    SOCKET Get() const override
+    {
+        assert(false && "Not implemented yet.");
+    }
+
+    SOCKET Release() override
+    {
+        assert(false && "Not implemented yet.");
+    }
+
+    void Reset() override
+    {
+        assert(false && "Not implemented yet.");
+    }
+
+    ssize_t Send(const void* data, size_t len, int flags) const override
+    {
+        constexpr std::array send_errnos{
+            EACCES,
+            EAGAIN,
+            EALREADY,
+            EBADF,
+            ECONNRESET,
+            EDESTADDRREQ,
+            EFAULT,
+            EINTR,
+            EINVAL,
+            EISCONN,
+            EMSGSIZE,
+            ENOBUFS,
+            ENOMEM,
+            ENOTCONN,
+            ENOTSOCK,
+            EOPNOTSUPP,
+            EPIPE,
+            EWOULDBLOCK,
+        };
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            return len;
+        }
+        const ssize_t r = m_fuzzed_data_provider.ConsumeIntegralInRange<ssize_t>(-1, len);
+        if (r == -1) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, send_errnos);
+        }
+        return r;
+    }
+
+    ssize_t Recv(void* buf, size_t len, int flags) const override
+    {
+        constexpr std::array recv_errnos{
+            EAGAIN,
+            EBADF,
+            ECONNREFUSED,
+            EFAULT,
+            EINTR,
+            EINVAL,
+            ENOMEM,
+            ENOTCONN,
+            ENOTSOCK,
+            EWOULDBLOCK,
+        };
+        assert(buf != nullptr || len == 0);
+        if (len == 0 || m_fuzzed_data_provider.ConsumeBool()) {
+            const ssize_t r = m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+            if (r == -1) {
+                SetFuzzedErrNo(m_fuzzed_data_provider, recv_errnos);
+            }
+            return r;
+        }
+        const std::vector<uint8_t> random_bytes = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(
+            m_fuzzed_data_provider.ConsumeBool() ?
+                len :
+                m_fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, len));
+        if (random_bytes.empty()) {
+            const ssize_t r = m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+            if (r == -1) {
+                SetFuzzedErrNo(m_fuzzed_data_provider, recv_errnos);
+            }
+            return r;
+        }
+        std::memcpy(buf, random_bytes.data(), random_bytes.size());
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            if (len > random_bytes.size()) {
+                std::memset((char*)buf + random_bytes.size(), 0, len - random_bytes.size());
```

Comment:
nit: the typecast `(char*)` in `memset()` should not be needed. Or if it is needed (why?), then it should also be needed for the `memcpy()` call a few lines earlier since both take `void*` argument.